lightningd/plugin.c

```diff
@@ -397,12 +412,67 @@ static void plugin_rpcmethod_destroy(struct json_command *cmd,
 	jsonrpc_command_remove(rpc, cmd->name);
 }
 
+static void plugin_rpcmethod_cb(const struct plugin_request *req,
+				struct plugin_rpc_request *rpc_req)
+{
+	// Parse
+	// Extract results or error
+	// Construct reply
+	// Return result with appropriate return code.
+}
+
 static void plugin_rpcmethod_dispatch(struct command *cmd, const char *buffer,
 				      const jsmntok_t *params)
 {
-	// FIXME: We could avoid parsing the buffer again if we were
-	// to also pass in the method name.
-	cmd->usage = "[params]";
+	const jsmntok_t *toks = params, *methtok, *idtok;
+	struct plugin_rpc_request *request;
+	struct plugins *plugins = cmd->ld->plugins;
+	struct plugin *plugin;
+
+	if (cmd->mode == CMD_USAGE) {
+		cmd->usage = "[params]";
+		return;
+	}
+
+	/* We're given the params, but we need to walk back to the
+	 * root object, so just walk backwards until the current
+	 * element has no parents, that's going to be the root
+	 * element. */
+	while (toks->parent != -1)
+		toks--;
+
+	methtok = json_get_member(buffer, toks, "method");
+	idtok = json_get_member(buffer, toks, "id");
+	/* We've parsed them before, these should not fail! */
+	assert(idtok != NULL && methtok != NULL);
+
+	request = tal(NULL, struct plugin_rpc_request);
+	request->method = tal_strndup(request, buffer + methtok->start,
+				      methtok->end - methtok->start);
+	request->id = tal_strndup(request, buffer + idtok->start,
+				      idtok->end - idtok->start);
+	request->params = tal_strndup(request, buffer + params->start,
+				      params->end - params->start);
+	request->plugin = NULL;
+	request->cmd = cmd;
+
+	/* Find the plugin that registered this RPC call */
+	list_for_each(&plugins->plugins, plugin, list) {
+		for (size_t i=0; i<tal_count(plugin->methods); i++) {
+			if (streq(request->method, plugin->methods[i])) {
+				request->plugin = plugin;
+				break;
```

Comment:
This break is a bit suboptimal, since it doesn't break out of the outer loop.

Cleanest is extract this code into 'find_plugin_for_cmd()', and have this `return plugin;`.

Reply:
Added fixup that uses a `goto` instead.